rllib/evaluation/worker_set.py

```diff
@@ -128,91 +130,201 @@ def __init__(
-                )
-            else:
-                spaces = None
+        # Try to add the actual env's obs/action spaces.
+        env_spaces = self.foreach_worker(
+            lambda worker: worker.foreach_env(
+                lambda env: (env.observation_space, env.action_space)
+            ),
+            remote_worker_indices=[0],
+            local_worker=False,
+        )
+        if env_spaces:
+            # env_spaces group spaces by environment then worker.
+            # So need to unpack thing twice.
+            spaces["__env__"] = env_spaces[0][0]
+
+        logger.info(
+            "Inferred observation/action spaces from remote "
+            f"worker (local worker has no env): {spaces}"
+        )
 
-            if local_worker:
-                self._local_worker = self._make_worker(
-                    cls=RolloutWorker,
-                    env_creator=env_creator,
-                    validate_env=validate_env,
-                    worker_index=0,
-                    num_workers=num_workers,
-                    config=self._local_config,
-                    spaces=spaces,
-                )
+        return spaces
 
+    @DeveloperAPI
     def local_worker(self) -> RolloutWorker:
         """Returns the local rollout worker."""
         return self._local_worker
 
+    @property
+    def _remote_workers(self) -> List[ActorHandle]:
+        """Returns the list of remote rollout workers."""
+        if log_once("worker_set__remote_workers"):
+            deprecation_warning(
+                old="_remote_workers",
+                help=(
+                    "Accessing remote workers directly through "
+                    "_remote_workers is strongly discouraged. "
+                    "Please try to use one of the foreach accessors "
+                    "that is fault tolerant. "
+                ),
+                error=False,
+            )
+        return list(self.__worker_manager.actors().values())
+
     def remote_workers(self) -> List[ActorHandle]:
-        """Returns a list of remote rollout workers."""
-        return self._remote_workers
+        """Returns the list of remote rollout workers."""
+        if log_once("worker_set_remote_workers"):
+            deprecation_warning(
+                old="remote_workers()",
+                help=(
+                    "Accessing the list of remote workers directly through "
+                    "remote_workers() is strongly discouraged. "
+                    "Please try to use one of the foreach accessors "
+                    "that is fault tolerant. "
+                ),
+                error=False,
+            )
+        return list(self.__worker_manager.actors().values())
 
+    @DeveloperAPI
+    def num_remote_workers(self) -> int:
+        return self.__worker_manager.num_actors()
+
+    @DeveloperAPI
+    def num_healthy_workers(self) -> int:
+        if self._local_worker and self.__worker_manager.num_healthy_actors() <= 0:
```

Comment:
Should we not assume that the local worker is always healthy? Otherwise, the entire algo would crash, no?
I would only return 1 here, if `num_rollout_workers` were configured as 0 in the first place by the user.

Reply:
ok how about this. simply do num_of_healthy_workers() plus 1 if self._local_worker is there?